### ISSUE TITLE
type Typevar bound interaction

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5581]
### Discussed in https://github.com/microsoft/pyright/discussions/5580

<div type='discussions-op-text'>

<sup>Originally posted by **hmc-cs-mdrissi** July 25, 2023</sup>
I'm confused by an interaction between bound=type TypeVar vs TypeVar using type[T].

```python
from typing import TypeVar

T = TypeVar("T")
T_type = TypeVar("T_type", bound=type)


def f1(x: type[T]) -> type[T]:
    return x


def f2(x: type) -> type:
    return x


def g(x: T_type) -> T_type:
    print(f1(x)) # Error produced here
    print(f2(x))
    return x

```

The error produced,

```
Argument of type "T_type@g" cannot be assigned to parameter "x" of type "type[T@f1]" in function "f1"
```

I don't follow why though. In g, x is type variable bound to type so it always a type. f1 should accept any type and return it back. f2 adjusts it type unsubscripted and is fine on next line.

Was unsure to open this as bug/question and if there's some missing rule between bound=type vs type[T] I'm not understanding.

edit:

```python
class Foo(Generic[T]):
    ...


T_foo = TypeVar("T_foo", bound=Foo[object])


def h1(x: Foo[T]) -> Foo[T]:
    return x


def h2(x: Foo[Any]) -> Foo[Any]:
    return x


def i(x: T_foo) -> T_foo:
    print(h1(x))
    print(h2(x))
    return x
```

This code with Foo instead of type does pass type checking and is similar in spirit to above but using different generic type then special type. I did Foo[Any] as I think of type equivalent type[Any].</div>

**Comment by erictraut** [comment]
This will be addressed in the next release.

Incidentally, I recommend against creating a type variable with a bound of `type`. It's almost always preferable to use `type[T]` rather than `T_type` where the latter has a bound of `type`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.320, which I just published. It will also be included in a future version of pylance.
